pina/pinn.py

```diff
@@ -155,59 +201,72 @@ def span_pts(self, *args, **kwargs):
                             argument['n'],
                             argument['mode'],
                             variables=argument['variables'])
-                        for argument in arguments)
+                            for argument in arguments)
             pts = merge_tensors(samples)
 
             # TODO
             # pts = pts.double()
-            pts = pts.to(dtype=self.dtype, device=self.device)
-            pts.requires_grad_(True)
-            pts.retain_grad()
-
             self.input_pts[location] = pts
 
     def train(self, stop=100, frequency_print=2, save_loss=1, trial=None):
 
         epoch = 0
+        data_loader = self._create_dataloader()
+        data_loader_loop = data_loader
```

Comment:
The dataset creation is just a view of the original tensors or do they implicitly perform new allocations?

Reply:
Can you be more specific?

Reply:
Is allocating twice the same data (input_pts and dataloader) or are they just references to same memory? Have you check that?

Reply:
from my understanding points are allocated in `input_pts` then in create dataloader we return a dict of iterables, each iterable takes a dataset as input (one for each key of the dict). A dataset has `_tensors` as private variable, so it is creating a shallow copy of `input_pts[key]`, which is needed for the `__get_item__` method.

Reply:
This is how I thought to handle different datasets (one for each key), for creating the dataloader and maintaining the same train routine. Notice that allocation is done on CPU, the tensors are moved to GPU in batches anyway. In practice  `input_pts` is always on CPU (even when trained on GPU).

Reply:
Umh, I'm lost since you're saying it's just a shallow copy, and few lines below you state that there are actually two copies of the same tensor.
From my knowledge, I'm not even sure it's a shallow copy. Maybe we can merge this PR and asap I gonna retouch the `span_pts` method in order to directly create the DataSet and always use such a data structure. Does it sound good for you?

Reply:
Ok for me, we can put it as a new issue and I can help in trying to fix it.